Add document how to integrate a custom Job with Kueue (#1016)
kubernetes-sigs/kueue: site/content/en/docs/tasks/integrate_a_custom_job.md

---
title: "Integrate a custom Job"

> date: 2022-07-25

Reviewer: ```suggestion
date: 2023-07-25
```

weight: 8
description: >
  Integrate a custom Job with Kueue.
---
# Integrate a custom Job with Kueue

This page shows how to integrate a custom Job with Kueue. 

> ## What we need to build

Reviewer: nit: I'm not sure `build` is the right word here. If the integration is part of the Kueue repo, then its build system will take care. I would suggest rather `provide` or `specify`.

Author: Got it. I think you're right.

> 1. Configration

Reviewer: ```suggestion
1. Configuration
```

2. Controller

> 3. Webhook

Reviewer: nit: it might be worth mentioning to say that the build system needs to be adjusted. The Makefile and the dependency.

## Configure your custom Job

> Add your framework name to `.integrations.frameworks` in `config/components/manager/controller_manager_config.yaml`

Reviewer: I think it would be nice to provide direct github links when providing the path. If we provide the path it means we need to maintain to keep in sync with code anyway.

## Build controller

> You can create the controller in `./pkg/controller/jobs/$YOUR_CUSTOM_JOB`.

Reviewer: nit: maybe we can rephrase slightly to say that you can provide the code of the controller under a dedicated directory, under the `./pkg/controller/jobs/` folder, so that we can provide a link.

1. Register a new framework
2. Add RBAC Authorization
    - Add RBAC Authorization for priorityclasses, events, workloads, resourceflavors and any resources you want.

> 3. Implement interfaces

Reviewer: I'm wondering if repeating the interface here is worth it. It adds maintenance cost to keep in sync. Maybe just a link to the interface would do.

    - `Object()`: returns the job instance.
    - `IsSuspended()`: returns whether the job is suspended or not.
    - `Suspend()`: will suspend the job.
    - `RunWithPodSetsInfo(nodeSelectors []PodSetInfo)`: will inject the node affinity and podSet counts extracting from workload to job and unsuspend it.
    - `RestorePodSetsInfo(nodeSelectors []PodSetInfo)`: will restore the original node affinity and podSet counts of the job.
    - `Finished()`: means whether the job is completed/failed or not, condition represents the workload finished condition.
    - `PodSets()`: will build workload podSets corresponding to the job.
    - `IsActive()`: returns true if there are any running pods.
    - `PodsReady()`: instructs whether job derived pods are all ready now.
    - `GetGVK()`: returns GVK (Group Version Kind) for the job.

## Build webhook

You can create the webhook in `./pkg/controller/jobs/$YOUR_CUSTOM_JOB`.

> You can learn how to create webhook in this [page](https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation.html). Also, you can see webhook [example](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobs/rayjob/rayjob_webhook.go) of RayJob with Kueue.

Reviewer: Additionally, we may suggest learning from  already existing integrations, and list them.

---
title: "Integrate a custom Job"
date: 2023-07-25
weight: 8
description: >
  Integrate a custom Job with Kueue.
---
# Integrate a custom Job with Kueue

This page shows how to integrate a custom Job with Kueue. 

## What you need to specify

1. Configuration
2. Controller
3. Webhook
4. Test files
5. Adjust build system

## Configure your custom Job

Add your framework name to `.integrations.frameworks` in [controller_manager_config.yaml](https://github.com/kubernetes-sigs/kueue/blob/main/config/components/manager/controller_manager_config.yaml)

## Build controller

Provide your code of the controller under a dedicated directory. This directory need to under the `./pkg/controller/jobs/` folder.

> 1. Register a new framework

Author: I added this because I wanted to mention that `jobframework.RegisterIntegration` is required to provide.

> 1. Register a new framework

Author: > Not sure what is meant here, I think the framework registration is what we do under `Add your framework name to `.integrations.frameworks` in [controller_manager_config.yaml](https://github.com/kubernetes-sigs/kueue/blob/main/config/components/manager/controller_manager_config.yaml)`. I would suggest to remove this point.

I added this because I wanted to mention that `jobframework.RegisterIntegration` is required to provide in `init()` in controller file.

> 1. Register a new framework

Reviewer: sgtm

> 2. Add RBAC Authorization

Author: This part I wanted to say that we need to add controller-tools RBAC markers like `// +kubebuilder:rbac:groups=scheduling.k8s.io,resources=priorityclasses,verbs=list;get;watch` in controller file.

> 2. Add RBAC Authorization

Author: > I think this point can be moved to configuration, as it is yaml configuration, maybe also provide with the link https://github.com/kubernetes-sigs/kueue/tree/main/config/components/rbac

This part I wanted to say that we need to add controller-tools RBAC markers like `// +kubebuilder:rbac:groups=scheduling.k8s.io,resources=priorityclasses,verbs=list;get;watch` in controller file.

> 2. Add RBAC Authorization

Reviewer: Got it, maybe we can clarify that by saying. Add RBAC authorization using [kubebuilder marker comments](https://book.kubebuilder.io/reference/markers/rbac.html)

> Provide your code of the controller under a dedicated directory. This directory need to under the `./pkg/controller/jobs/` folder.

Reviewer 2: ```suggestion
Add a new folder in `./pkg/controller/jobs/` to host the implementation of the integration.
```

> 1. Register the new framework

Reviewer 2: where?

Author: You mean I need to mention where need to register the new framework?

Author: > where?

You mean I need to mention where need to register the new framework?

Reviewer 2: yes please

2. Add RBAC Authorization using [kubebuilder marker comments](https://book.kubebuilder.io/reference/markers/rbac.html)
    - Add RBAC Authorization for priorityclasses, events, workloads, resourceflavors and any resources you want.

> 3. Implement the `GenericJob` interface, and potentially other interfaces in [here](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobframework/interface.go).

Reviewer 2: ```suggestion
3. Implement the `GenericJob` interface, and other optional [interfaces defined by the framework](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobframework/interface.go).
```

>     - You can click this [link](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobframework/interface.go) to check which interfaces you need to implement.

Reviewer 2: this is already mentioned in the line before.

## Webhook

> Create your webhook file in your dedicated director.

Reviewer 2: ```suggestion
Create a webhook file in the dedicated directory for the integration.
```

> You can learn how to create webhook in this [page](https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation.html).

Reviewer 2: Also mention that the framework provides some validation functions for common labels and annotations.

Make sure that your custom job has **suspend** abilities. Kueue will only schedule suspended job.

## Add test files
Create test files which help test your custom job's controller and webhook.

You can check the sample test files in [completed integrations](#completed-integrations) to learn how to implement them.

## Adjust build system

> Update [Makefile](https://github.com/kubernetes-sigs/kueue/blob/main/Makefile) for testing.

Reviewer: Maybe first mention to add required dependencies to compile the controller and webhook code. We may mention that for example, using `go get github.com/kubeflow/mpi-operator@0.4.0`.

> Your webhook should have ability to suspend jobs.

Reviewer 2: I think there should be a section for pre-requisites, indicating that the Job CRD should have a suspend-like field, and you can point to the Job documentation as an example: https://kubernetes.io/docs/concepts/workloads/controllers/job/#suspending-a-job

> ## Adjust build system

Reviewer 2: We should mention that integrations can also live in an external repo, while still using the same integration framework.

Author: I just put this point below `## Adjust build system` ? Or other places

Author: > We should mention that integrations can also live in an external repo, while still using the same integration framework.

I just put this point below ## Adjust build system ? Or other places

---
title: "Integrate a custom Job"
date: 2023-07-25
weight: 8
description: >
  Integrate a custom Job with Kueue.
---
# Integrate a custom Job with Kueue

> Kueue integrates with a couple of jobs, including batch Job, MPI Job, Ray Job or JobSet. 

Reviewer 3: ```suggestion
Kueue integrates with a couple of jobs, including Kubernetes batch Job, MPIJob, RayJob or JobSet. 
```

If there is another Job, which you would like supported, please consider contributing. 
This page shows how to integrate a custom Job with Kueue.

## What you need to specify

1. Configuration
2. Controller
3. Webhook
4. Adjust build system
5. Test files

> ## Configure your custom Job

Reviewer 3: Maybe better to make it a `###Configure your custom Job` and make the name consist with the names under `What you need to specify`.

Add your framework name to `.integrations.frameworks` in [controller_manager_config.yaml](https://github.com/kubernetes-sigs/kueue/blob/main/config/components/manager/controller_manager_config.yaml)

## Controller

Provide your code of the controller under a dedicated directory. This directory need to under the `./pkg/controller/jobs/` folder.

1. Register the new framework
2. Add RBAC Authorization using [kubebuilder marker comments](https://book.kubebuilder.io/reference/markers/rbac.html)
    - Add RBAC Authorization for priorityclasses, events, workloads, resourceflavors and any resources you want.
3. Implement the `GenericJob` interface, and potentially other interfaces in [here](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobframework/interface.go).
    - You can click this [link](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobframework/interface.go) to check which interfaces you need to implement.

## Webhook

Create your webhook file in your dedicated director.

> You can learn how to create webhook in this [page](https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation.html).

Reviewer 3: Maybe we should notice here that: suspend your job in webhook, also since `suspend` capacity is critical, we'd better highlight that at the beginning.

## Adjust build system
1. Add required dependencies to compile the controller and webhook code. For example, using `go get github.com/kubeflow/mpi-operator@0.4.0`.
2. Update [Makefile](https://github.com/kubernetes-sigs/kueue/blob/main/Makefile) for testing.
   - Add commands which copy the crd of your custom job to the Kueue project.
   - Add your custom job operator crd dependencies into `test-integration`.

> ## Add test files

Reviewer 3: Can we remove this section, I think this is not that important just a best practice in developing. Hope to make this document as concise as possible.

---
title: "Integrate a custom Job"
date: 2023-07-25
weight: 8
description: >
  Integrate a custom Job with Kueue.
---
# Integrate a custom Job with Kueue

Kueue integrates with a couple of jobs, including batch Job, MPI Job, Ray Job or JobSet. 
If there is another Job, which you would like supported, please consider contributing. 
This page shows how to integrate a custom Job with Kueue.

## What you need to specify

1. Configuration
2. Controller
3. Webhook
4. Adjust build system
5. Test files

## Configure your custom Job

Add your framework name to `.integrations.frameworks` in [controller_manager_config.yaml](https://github.com/kubernetes-sigs/kueue/blob/main/config/components/manager/controller_manager_config.yaml)

## Controller

Provide your code of the controller under a dedicated directory. This directory need to under the `./pkg/controller/jobs/` folder.

> 1. Register a new framework

Reviewer: ```suggestion
1. Register the new framework
```

2. Add RBAC Authorization
    - Add RBAC Authorization for priorityclasses, events, workloads, resourceflavors and any resources you want.

> 3. Implement interfaces

Reviewer: I think the `GenericJob` is particularly important so it should be called here. Maybe
```
3. Implement the `GenericJob` interface, and potentially other interfaces in [here](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobframework/interface.go).
```

---
title: "Integrate a custom Job with Kueue"
date: 2023-07-25
weight: 8
description: >
  Integrate a custom Job with Kueue.
---

Kueue integrates with a couple of Jobs, including Kubernetes batch Job, MPIJob, RayJob or JobSet.  

> If there is another Job, which you would like supported, please consider contributing. 

Reviewer 2: ```suggestion
There are two options for integrating a Job-like CRD with Kueue:
- As part of the Kueue repository.
- Writing an external controller
This guide focuses on the first approach.
If there is a widely used Job which you would like supported in Kueue, please consider contributing. 
```

---

> title: "Integrate a custom Job"

Reviewer: ```suggestion
title: "Integrate a custom Job with Kueue"
```

date: 2023-07-25
weight: 8
description: >
  Integrate a custom Job with Kueue.
---

> # Integrate a custom Job with Kueue

Reviewer: No need to repeat as the title is displayed based on `title` field.

> Kueue integrates with a couple of jobs, including Kubernetes batch Job, MPIJob, RayJob or JobSet.  

Reviewer: ```suggestion
Kueue integrates with a couple of Jobs, including Kubernetes batch Job, MPIJob, RayJob or JobSet.  
```

4. Adjust build system

## Configuration

Add your framework name to `.integrations.frameworks` in [controller_manager_config.yaml](https://github.com/kubernetes-sigs/kueue/blob/main/config/components/manager/controller_manager_config.yaml)

## Controller

Add a new folder in `./pkg/controller/jobs/` to host the implementation of the integration.

1. Register the new framework in `func Init()`
2. Add RBAC Authorization using [kubebuilder marker comments](https://book.kubebuilder.io/reference/markers/rbac.html)
    - Add RBAC Authorization for priorityclasses, events, workloads, resourceflavors and any resources you want.
3. Implement the `GenericJob` interface, and other optional [interfaces defined by the framework](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobframework/interface.go).

## Webhook

Create a webhook file in the dedicated directory for the integration.

You can learn how to create webhook in this [page](https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation.html).

The framework provides some validation functions for common labels and annotations.

Your webhook should have ability to suspend jobs.

## Adjust build system

> Integrations can also live in an external repo, while still using the same integration framework.

Reviewer: +1, I think it is ok to be mention, but be brief about integrations in external repos.

Later when, we have examples of integrations in external repos we may expand the section with examples and a guide.

Reviewer 2: This looks like the wrong place to say this. See my comment in line 10.
